### PR TITLE
Feature RegUser application

### PR DESCRIPTION
**Pull request by TrupeshKumarPatel**

### CODE REVIEW

File: roles/ood_user_reg/tasks/main.yml [hunk 1]
```diff
@@ -66,3 +66,29 @@
...
   ignore_errors: yes
+
+- name: Create RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo id -u {{ RegUser_app_user }} &>/dev/null || sudo useradd -c "{{ RegUser_app_user_full_name }}" {{ RegUser_app_user }}'
+  become_user: vagrant
+
+- name : delay
+  command: sleep 2
+
+- name: Change passwd RegUser - {{ RegUser_app_user }} on OHPC
```
Comment (jprorama) [inline roles/ood_user_reg/tasks/main.yml]: Do we really need a password?  We really shouldn't set this for a service account.

File: roles/ood_user_reg/tasks/main.yml [hunk 2]
```diff
@@ -66,3 +66,29 @@
...
+  shell: ssh ohpc 'sudo echo "{{ RegUser_app_user_passwd }}" | sudo passwd --stdin "{{ RegUser_app_user }}"'
+  become_user: vagrant
+
+- name: Create Cluster keys for RegUser - {{ RegUser_app_user }} on OHPC
```
Comment (jprorama) [inline roles/ood_user_reg/tasks/main.yml]: Separate to ohpc role.

File: roles/ood_user_reg/tasks/main.yml [hunk 3]
```diff
@@ -66,3 +66,29 @@
 - name: Build the updated Apache config
   command: /opt/ood/ood-portal-generator/sbin/update_ood_portal
   ignore_errors: yes
+
+- name: Create RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo id -u {{ RegUser_app_user }} &>/dev/null || sudo useradd -c "{{ RegUser_app_user_full_name }}" {{ RegUser_app_user }}'
```
Comment (jprorama) [inline roles/ood_user_reg/tasks/main.yml]: Make sure to use the ansible user create module.

https://docs.ansible.com/ansible/latest/modules/user_module.html

File: ohpc.yaml [hunk 4]
```diff
@@ -11,3 +11,5 @@
     - { name: 'ohpc_jupyter', tags: 'ohpc_jupyter', when: jupyter_provision}
     - { name: 'ohpc_matlab', tags: 'ohpc_matlab', when: matlab_provision }
     - { name: 'ohpc_add_rstudio', tags: 'ohpc_add_rstudio', when: rstudio_provision }
+    - { name: 'reg_user_create_scripts', tags: 'reg_user_create_scripts', when: enable_user_create_scripts}
+
```
Comment (jprorama) [inline ohpc.yaml]: Add a role for the registration service account user on ohpc: user create, key create, sudoers rule tasks.

File: ood.yaml [hunk 5]
```diff
@@ -3,8 +3,8 @@
   roles:
     - { name: 'prep_ood', tags: 'prep_ood' }
     - { name: 'ood', tags: 'ood' }
-    - { name: 'ood_user_reg', tags: 'ood_user_reg', when: enable_user_reg }
     - { name: 'warewulf_sync', tags: 'warewulf_sync' }
+    - { name: 'ood_user_reg', tags: 'ood_user_reg', when: enable_user_reg }
```
Comment (jprorama) [inline ood.yaml]: Clean this up.  Shouldn't need to change the order of this roles.

File: roles/ood_user_reg/tasks/main.yml [hunk 6]
```diff
@@ -66,3 +66,29 @@
...
+
+- name: Change passwd RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo echo "{{ RegUser_app_user_passwd }}" | sudo passwd --stdin "{{ RegUser_app_user }}"'
+  become_user: vagrant
+
+- name: Create Cluster keys for RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo -s su - reggie -c "sh /usr/bin/cluster-env"'
+  become_user: vagrant
+
+- name: Get UID of RegUser - {{ RegUser_app_user }} from OHPC
```
Comment (jprorama) [inline roles/ood_user_reg/tasks/main.yml]: Do this on ood via inspection of uid, gid owner of /home/reggie.  Then can use those uid, gid in the user module create request within ansible task on ood role.

File: roles/ood_user_reg/tasks/main.yml [hunk 7]
```diff
@@ -66,3 +66,29 @@
...
+
+- name : delay
+  command: sleep 2
+
+- name: Change passwd RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo echo "{{ RegUser_app_user_passwd }}" | sudo passwd --stdin "{{ RegUser_app_user }}"'
+  become_user: vagrant
+
+- name: Create Cluster keys for RegUser - {{ RegUser_app_user }} on OHPC
+  shell: ssh ohpc 'sudo -s su - reggie -c "sh /usr/bin/cluster-env"'
```
Comment (jprorama) [inline roles/ood_user_reg/tasks/main.yml]: Should be able to trim this to just `su - reggie -c "/usr/bin/cluster-env"`

Comment (diedpigs) [reply roles/ood_user_reg/tasks/main.yml]: i thought we are just logging and exit out? e.g. `su - reggie exit` ? Do we have to execute the script explicitly?

File: roles/ood_user_reg/tasks/main.yml (hunk 3 above)
Comment (flakrat) [inline roles/ood_user_reg/tasks/main.yml]: NOTE: It's not necessary to use `sudo` for `id -u {{ RegUser_app_user }}`.

`sudo` should only be used when required, as it's not good security practice to elevate privileges when unnecessary.

Comment (flakrat) [reply roles/ood_user_reg/tasks/main.yml]: Additionally, this user should probably be created as a `system` user, i.e. `system: yes` as better clarifies the role of the user and will assign it an id from the system user range vs the higher range used by normal users.

File: roles/ood_user_reg/tasks/main.yml (hunk 7 above)
Comment (TrupeshKumarPatel) [reply roles/ood_user_reg/tasks/main.yml]: yeah but i think that did not work.

File: roles/ood_user_reg/tasks/main.yml (hunk 3 above)
Comment (TrupeshKumarPatel) [reply roles/ood_user_reg/tasks/main.yml]: But i think we do not want this user as `system` user, right?

**Pull request merged**